vfs/httpfs/union/union.go

```diff
@@ -0,0 +1,128 @@
+package union
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/shurcooL/go/vfs/httpfs/vfsutil"
+)
+
+// New creates an union filesystem with the provided mapping of mount points to filesystems.
+//
+// Each mount point must be of form "/mydir". It must start with a '/', and contain a single directory name.
+func New(mapping map[string]http.FileSystem) http.FileSystem {
+	u := &unionFS{
+		ns: make(map[string]http.FileSystem),
+		root: &dirInfo{
+			name:    "/",
+			entries: []os.FileInfo{}, // Not nil.
+			modTime: time.Time{},     // An empty union filesystem has never been modified.
+		},
+	}
+	for mountPoint, fs := range mapping {
+		u.bind(mountPoint, fs)
+	}
+	return u
+}
+
+type unionFS struct {
+	ns   map[string]http.FileSystem // Key is mount point, e.g., "/mydir".
+	root *dirInfo
+}
+
+// bind mounts fs at mountPoint.
+// mountPoint must be of form "/mydir". It must start with a '/', and contain a single directory name.
+func (u *unionFS) bind(mountPoint string, fs http.FileSystem) {
+	fi, err := vfsutil.Stat(fs, "/")
+	if err != nil {
+		log.Fatalln("can't stat root directory of provided filesystem:", err)
+	}
+	modTime := fi.ModTime()
+
+	u.ns[mountPoint] = fs
+	u.root.entries = append(u.root.entries, &dirInfo{
+		name:    mountPoint[1:],
+		entries: nil, // entries is nil, not needed because this acts as a os.FileInfo only, not an openable dir.
+		modTime: modTime,
+	})
+
+	if modTime.After(u.root.modTime) {
+		u.root.modTime = modTime
+	}
+}
+
+// Open opens the named file.
+func (u *unionFS) Open(path string) (http.File, error) {
+	// TODO: Maybe clean path?
+	if path == "/" {
+		return &dir{
+			dirInfo: u.root,
+		}, nil
+	}
+	for prefix, fs := range u.ns {
+		if path == prefix || strings.HasPrefix(path, prefix+"/") {
+			innerPath := path[len(prefix):]
+			if innerPath == "" {
+				innerPath = "/"
+			}
+			return fs.Open(innerPath)
+		}
+	}
+	return nil, &os.PathError{Op: "open", Path: path, Err: os.ErrNotExist}
+}
+
+// dirInfo is a static definition of a directory.
+type dirInfo struct {
+	name    string
+	entries []os.FileInfo // Not nil.
+	modTime time.Time
+}
+
+func (d *dirInfo) Read([]byte) (int, error) {
+	return 0, fmt.Errorf("cannot Read from directory %s", d.name)
+}
+func (d *dirInfo) Close() error               { return nil }
+func (d *dirInfo) Stat() (os.FileInfo, error) { return d, nil }
+
+func (d *dirInfo) Name() string       { return d.name }
+func (d *dirInfo) Size() int64        { return 0 }
+func (d *dirInfo) Mode() os.FileMode  { return 0755 | os.ModeDir }
+func (d *dirInfo) ModTime() time.Time { return d.modTime }
+func (d *dirInfo) IsDir() bool        { return true }
+func (d *dirInfo) Sys() interface{}   { return nil }
+
+// dir is an opened dir instance.
+type dir struct {
+	*dirInfo
+	pending []os.FileInfo
+}
+
+func (d *dir) Seek(offset int64, whence int) (int64, error) {
+	if offset == 0 && whence == os.SEEK_SET {
+		d.pending = nil
```

Comment:
Actually, most other methods are safe for concurrent use. For example, you can open multiple files (same and different paths) concurrently. But for each opened file, you can't call methods that modify state (for example, Read or Seek) from multiple goroutines. It's possible to open the same file multiple times and access each one independently and concurrently.